File: LeetCode/20 - Valid Parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        for index in range(len(s)):
            stackLen = len(stack)
            if (stackLen > 0 and stack[stackLen - 1] == self.opposite(s[index])):
                stack.pop()
            else:
                stack.append(s[index])

        return len(stack) == 0

    def opposite(self, char):
        if char == ")":
            return "("
        elif char == "}":
            return "{"
        elif char == "]":
            return "["

File: LeetCode/20 - Valid Parentheses.ts
function isValid(s: string): boolean {
  function isMatchingParanthesis(lhs, rhs): boolean {
    if (lhs === "(" && rhs === ")") return true
    if (lhs === "[" && rhs === "]") return true
    if (lhs === "{" && rhs === "}") return true
    return false
  }

  const stack: number[] = []
  for (let index = 0; index < s.length; index += 1) {
    if (isMatchingParanthesis(stack.at(-1), s[index])) {
      stack.pop()
    } else {
      stack.push(s[index])
    }
  }

  if (stack.length === 0) return true
  return false
}
